### PR TITLE
feat: #1185 Add a 'Ping' feature to the Webhook Subscriptions

### PR DESCRIPTION
**Pull request by ant066**

### CODE REVIEW

File: packages/marketplace/src/actions/developer.ts [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
 import { actionCreator } from '../utils/actions'
 import ActionTypes from '../constants/action-types'
-import { DeveloperItem, DeveloperRequestParams, Billing, MonthlyBilling } from '../reducers/developer'
+import {
+  DeveloperItem,
+  DeveloperRequestParams,
+  Billing,
+  MonthlyBilling,
+  WebhookPingTestStatus,
+} from '../reducers/developer'
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/actions/developer.ts:10]: suggest to change to `@/reducers/developer`

File: packages/marketplace/src/components/pages/developer-webhooks.tsx [hunk 2]
```diff
@@ -214,20 +230,26 @@ export const DeveloperWebhooks = ({
...
               scrollable
               columns={columns}
-              data={getTableTopicsData({ subscriptions, handleOpenEditModal, topics })}
+              data={getTableTopicsData({ subscriptions, handleOpenEditModal, topics, handleOpenTestModal })}
               loading={false}
             />
           ) : null}
         </FormSection>
       </FlexContainerResponsive>
       <WebhookEditModal
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/pages/developer-webhooks.tsx:239]: suggest to change move this one to logic
```
const isShowDetailModal = modalType === MODAL_TYPE.EDIT || modalType === MODAL_TYPE.CREATE
...
isShowDetailModal && <WebhookEditModal
....
````

File: packages/marketplace/src/components/pages/developer-webhooks.tsx [hunk 3]
```diff
@@ -214,20 +230,26 @@ export const DeveloperWebhooks = ({
...
-        visible={!!modalType}
+        visible={modalType === MODAL_TYPE.EDIT || modalType === MODAL_TYPE.CREATE}
         isUpdate={modalType === MODAL_TYPE.EDIT}
         appId={applicationId}
         webhookId={webhookId}
         afterClose={afterClose}
         closeModal={handleCloseModal}
       />
+      <WebhookTestModal
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/pages/developer-webhooks.tsx:247]: suggest to change
```
const isShowTestModal = modalType === MODAL_TYPE.TEST
...
isShowTestModal && <WebhookTestModal
...
```

File: packages/marketplace/src/components/pages/developer-webhooks.tsx [hunk 4]
```diff
@@ -214,20 +230,26 @@ export const DeveloperWebhooks = ({
...
+        visible={modalType === MODAL_TYPE.TEST}
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/pages/developer-webhooks.tsx:248]: `visible={isShowTestModal}`

File: packages/marketplace/src/components/pages/developer-webhooks.tsx [hunk 5]
```diff
@@ -214,20 +230,26 @@ export const DeveloperWebhooks = ({
...
-              data={getTableTopicsData({ subscriptions, handleOpenEditModal, topics })}
+              data={getTableTopicsData({ subscriptions, handleOpenEditModal, topics, handleOpenTestModal })}
               loading={false}
             />
           ) : null}
         </FormSection>
       </FlexContainerResponsive>
       <WebhookEditModal
-        visible={!!modalType}
+        visible={modalType === MODAL_TYPE.EDIT || modalType === MODAL_TYPE.CREATE}
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/pages/developer-webhooks.tsx:240]: `visible={isShowDetailModal}`

File: packages/marketplace/src/components/ui/webhook-test-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  closeModal?: () => void
+}
+
+export interface GenerateTopicOptions {
+  topics: TopicItem[]
+  subcriptionTopicIds: string[]
+}
+export const generateTopicOptions = ({ topics, subcriptionTopicIds }: GenerateTopicOptions): SelectOption[] => {
+  return subcriptionTopicIds.map(topicId => {
+    const topicData: TopicItem | undefined = topics.find(topic => topic.id === topicId)
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/ui/webhook-test-modal.tsx:38]: `    const topicData: TopicItem | undefined = topics.find((topic: TopicItem) => topic.id === topicId)`

File: packages/marketplace/src/components/ui/webhook-test-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    <Modal visible={visible} renderChildren afterClose={afterClose}>
+      {loading ? (
+        <Loader />
+      ) : (
+        <Formik initialValues={initFormValues} onSubmit={onSubmit}>
+          <Form>
+            <ModalHeader title="Test Webhook Subscription" />
+            <ModalBody
+              body={
+                <>
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/ui/webhook-test-modal.tsx:95]: Suggest to move this content to new function to look cleaner

File: packages/marketplace/src/components/ui/webhook-test-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                    name="topicId"
+                    labelText="Subscription Topics"
+                    options={topicOptions}
+                    required
+                  />
+                </>
+              }
+            />
+            <ModalFooter
+              footerItems={
```
Comment (tanphamhaiduong) [inline packages/marketplace/src/components/ui/webhook-test-modal.tsx:110]: Suggest to move this content to new function to look cleaner

File: packages/marketplace/src/services/__tests__/billings.ts [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
+import * as Element from '@reapit/elements'
 
 describe('should fetchMonthlyBilling run correctly', () => {
   it('fetcher should be called', () => {
     const applicationId = ['applicationId']
     const month = '2020-05'
     const headers = generateHeader(window.reapit.config.marketplaceApiKey)
-    const url = `${URLS.trafficEvents}/billing/${month}?${setQueryParams({ applicationId })}`
+    const url = `${URLS.trafficEvents}/billing/${month}?${Element.setQueryParams({ applicationId })}`
+    const fetcher = spyOn(Element, 'fetcher')
```
Comment (willmcvay) [inline packages/marketplace/src/services/__tests__/billings.ts:11]: You can just `spyOn(window, 'fetch')` as we mock the fetch object on jest.setup. Or, you can just jest.mock('@reapit/elements') and will pick up the fetcher mock in the Elements mock folder

File: packages/marketplace/src/services/__tests__/subcriptions.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import * as Element from '@reapit/elements'
+import * as API from '@/constants/api'
+
+describe('should fetchMonthlyBilling run correctly', () => {
+  it('fetcher should be called', () => {
+    const id = 'applicationId'
+    const topicId = 'topicId'
+
+    const url = `${API.URLS.webhook}/subscriptions/${id}/ping`
+    const fetcher = spyOn(Element, 'fetcher')
```
Comment (willmcvay) [inline packages/marketplace/src/services/__tests__/subcriptions.ts:11]: As above, see mocking fetch

File: packages/marketplace/src/services/__tests__/topics.ts [hunk 11]
```diff
@@ -1,16 +1,27 @@
...
-      .then(() => {
-        expect(initAuthorizedRequestHeaders).toBeCalled()
-      })
-      .then(() => {
-        expect(fetcher).toBeCalledWith()
+    const url = `${API.URLS.webhookTopics}?${Element.setQueryParams({ applicationId })}`
+    const method = 'GET'
+    const api = window.reapit.config.platformApiUrl
+
+    const fetcher = spyOn(Element, 'fetcher')
```
Comment (willmcvay) [inline packages/marketplace/src/services/__tests__/topics.ts:12]: And again, see mocking fetch

File: packages/marketplace/src/services/__tests__/billings.ts (hunk 9 above)
Comment (ant066) [reply packages/marketplace/src/services/__tests__/billings.ts]: I updated it. you can review it again

**Pull request merged**